extmod/uasyncio/core.py

```diff
@@ -145,26 +146,32 @@ def create_task(coro):
     _task_queue.push(t)
     return t
 
-
 # Keep scheduling tasks until there are none left to schedule
 def run_until_complete(main_task=None):
     global cur_task
     excs_all = (CancelledError, Exception)  # To prevent heap allocation in loop
     excs_stop = (CancelledError, StopIteration)  # To prevent heap allocation in loop
+    _io_queue.wait_io_event(0)
     while True:
-        # Wait until the head of _task_queue is ready to run
-        dt = 1
-        while dt > 0:
-            dt = -1
```

Comment:
I'm pretty sure this `while dt > 0` loop is there for a reason... what if `wait_io_event()` returns early yet does not queue anything (is that possible?), then the next available task on the run queue will be scheduled early.

Reply:
I think that would be a bug in the `poller::ipoll` method, but I certainly can't fault your careful approach. Thank you for bringing it up.

It took me a while to understand how the code works the way you describe:
- _(assuming there is one task waiting with a ph_key delta of 1000ms)_
- dt is set to 1
- LOOP (dt == 1)
- Check if a task is waiting
- Calculate dt (will be set to 1000)
- Run `wait_io_event` with dt := 1000
- LOOP (dt == 1000)
- Check if a task is waiting
- Calculate dt (will be ideally set to <= 0, but could be a positive number)
- Run `wait_io_event` with dt := 0
- EXIT LOOP (dt <= 0)

When the logic I'm proposing:
- _(assuming there is one task waiting with a ph_key delta of 1000ms)_
- LOOP
- Check if a task is waiting
- Calculate dt (will be set to 1000)
- Run `wait_io_event` with dt := 1000
- LOOP
- Calculate dt (will ideally be set to <= 0, but could be a positive number))
- BREAK if dt <= 0 _(any immediately pending IO will be run after the ready task)_

This means that `wait_io_event` is run only once per cycle rather than twice (unless it returns early).

I also took the liberty of making slight optimizations to the `wait_io_queue` method to reduce some opcode overhead by avoiding lookups and calculations on the POLLIN and POLLOUT constants as well as unpacking the map entry instead of using item lookups to fetch [0] and [1] repeatedly.

Here's a simple benchmark script I used with results:
```python
import asyncio

counter = 0
async def count():
    global counter
    while True:
        await asyncio.sleep_ms(0)
        counter += 1

try:
    asyncio.run(asyncio.wait_for(count(), timeout=2))
finally:
    print(counter, 2e6/counter)
```

(unix) Apple M2 -------
WITH PATCH:
| schedules | us/schedule |
|-|-|
| 236425 | 8.459 |

ON v1.23.0-preview.203.gd712feb68
| schedules | us/schedule |
|-|-|
| 219436 | 9.114 |

(rp2) RP2040 (@125MHz) ---------
WITH PATCH
| schedules | us/schedule |
|-|-|
| 12609 | 158.6169 |
| 12611 | 158.5917 |

ON v1.22.2
| schedules | us/schedule |
|-|-|
| 6442 | 310.4626 |

The results on the RP2 are a bit dramatic. I think we hit some magic alignment victory or something. But I wanted to demonstrate that it's not slower than the baseline.

Cheers,